extension/healthcheckv2extension/internal/status/aggregator.go

```diff
@@ -0,0 +1,225 @@
+	extsIDMap = map[component.ID]struct{}{extsID: {}}
+)
+
+// Note: this interface had to be introduced because we need to be able to rewrite the
+// timestamps of some events during aggregation. The implementation in core doesn't currently
+// allow this, but this interface provides a workaround.
+type Event interface {
+	Status() component.Status
+	Err() error
+	Timestamp() time.Time
+}
+
+// Scope refers to a part of an AggregateStatus. The zero-value, aka ScopeAll,
+// refers to the entire AggregateStatus. ScopeExtensions refers to the extensions
+// subtree, and any other value refers to a pipeline subtree.
+type Scope string
+
+const (
+	ScopeAll        Scope  = ""
+	ScopeExtensions Scope  = "extensions"
+	pipelinePrefix  string = "pipeline:"
+)
+
+func (s Scope) toKey() string {
+	if s == ScopeAll || s == ScopeExtensions {
+		return string(s)
+	}
+	return pipelinePrefix + string(s)
+}
+
+type Verbosity bool
+
+const (
+	Verbose Verbosity = true
+	Concise           = false
+)
+
+// AggregateStatus contains a map of child AggregateStatuses and an embedded Event.
+// It can be used to represent a single, top-level status when the ComponentStatusMap
+// is empty, or a nested structure when map is non-empty.
+type AggregateStatus struct {
+	Event
+
+	ComponentStatusMap map[string]*AggregateStatus
+}
+
+func (a *AggregateStatus) clone(verbosity Verbosity) *AggregateStatus {
+	st := &AggregateStatus{
+		Event: a.Event,
+	}
+
+	if verbosity == Verbose && len(a.ComponentStatusMap) > 0 {
+		st.ComponentStatusMap = make(map[string]*AggregateStatus, len(a.ComponentStatusMap))
+		for k, cs := range a.ComponentStatusMap {
+			st.ComponentStatusMap[k] = cs.clone(verbosity)
+		}
+	}
+
+	return st
+}
+
+type subscription struct {
+	statusCh  chan *AggregateStatus
+	verbosity Verbosity
+}
+
+// Aggregator records individual status events for components and aggregates statuses for the
+// pipelines they belong to and the collector overall.
+type Aggregator struct {
+	mu              sync.RWMutex
+	aggregateStatus *AggregateStatus
+	subscriptions   map[string][]*subscription
+	aggregationFunc aggregationFunc
+}
+
+// NewAggregator returns a *status.Aggregator.
+func NewAggregator(errPriority ErrorPriority) *Aggregator {
+	return &Aggregator{
+		aggregateStatus: &AggregateStatus{
+			Event:              &component.StatusEvent{},
+			ComponentStatusMap: make(map[string]*AggregateStatus),
+		},
+		subscriptions:   make(map[string][]*subscription),
+		aggregationFunc: newAggregationFunc(errPriority),
+	}
+}
+
+// AggregateStatus returns an *AggregateStatus for the given scope. The scope can be the collector
+// overall (ScopeAll), extensions (ScopeExtensions), or a pipeline by name. Detail specifies whether
+// or not subtrees should be returned with the *AggregateStatus. The boolean return value indicates
+// whether or not the scope was found.
+func (a *Aggregator) AggregateStatus(scope Scope, verbosity Verbosity) (*AggregateStatus, bool) {
+	a.mu.Lock()
```

Comment:
RLock/RUnlock? I think you are only reading values here, right? Or is the clone operation changing `a` as well?